referrer-policy/generic/referrer-policy-test-case.sub.js

```diff
@@ -144,40 +138,64 @@ function ReferrerPolicyTestCase(scenario, testDescription, sanityChecker) {
     // `Referer` headers with length over 4k are culled down to an origin, so, let's test around
     // that boundary for tests that would otherwise return the complete URL.
     if (scenario.referrer_url == "stripped-referrer") {
-      promise_test(_ => {
+      promise_test(t => {
         history.pushState(null, null, "/");
         history.replaceState(null, null, "A".repeat(4096 - location.href.length - 1));
         const expectedReferrer = location.href;
         // Ensure that we don't load the same URL as the previous test.
         subresource.url += "&-1";
         return invokeRequest(subresource, [])
           .then(result => checkResult(location.href, result))
-          .then(_ => history.back())
-          .then(asyncResolve);
+          .finally(_ => {
+              history.back();
+              return new Promise(resolve => {
+                  window.addEventListener('popstate', resolve, {once: true});
+                  // Workaround for Safari: Waiting for popstate events causes
+                  // timeout in a-tag tests. To avoid timeout, we anyway resolve
+                  // the promise.
+                  if (scenario.subresource === 'a-tag') {
+                    t.step_timeout(resolve, 1000);
+                  }
+                });
+            });
```

Comment:
Could you please make this a helper helper and not copy-paste.

Reply:
Done.